MIPs/mip-3.md

```diff
@@ -0,0 +1,186 @@
+---
+MIP: 3
+Title: Implement `wallet_swapAsset` to support Dapps to swap
+Status: Draft
+Stability: Experimental
+discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions
+Author(s): Tomás Santos (@tommasini)
+Type: Community
+Created: 2023-11-10
+---
+
+## Summary
+
+This proposal introduces a new JSON RPC method `wallet_swapAsset` that allows dApps to request MetaMask to perform a token swap operation. The method accepts parameters for the source and destination tokens and initiates a token swap operation.
+
+## Motivation
+
+This proposal aims to solve a key pain point for dApp developers and users - the inability to easily swap tokens within the dApp interface.
+
+Currently, if a dApp requires users to swap from token A to token B to perform a specific operation, the user has to exit the dApp, go to MetaMask or third parties, search for the token the user needs and execute the swap, then return to the dApp. This disjointed workflow results in a poor user experience.
+
+Implementing `wallet_swapAsset` would allow the entire token swap process to occur seamlessly within the dApp without requiring external integration. This provides the following benefits:
+
+- Improved usability - Users can get the right tokens for dApp operations in fewer steps.
+- Better dApp integration - dApps can build and control the swap experience natively with just a few lines.
+- Increased adoption - More intuitive swap workflow lowers barrier to dApp usage.
+- Interoperability - Standard method works across different wallets.
+- Increase revenue - Easier adoption means more users will use MetaMask swaps.
+
+Comparing with other alternative, implementing it using snaps, would create more friction and less adoption since the users would need to install that specific snap to be able to perform a swap action. The goal is to reduce friction to swaps.
```

Comment:
```suggestion
Introducing this API directly in the wallet is preferable when compared with the alternative of using snaps, as the latter would create more friction and less adoption since users would need to install that specific snap to be able to perform a swap action. The goal is to reduce friction to swaps.
```